### ISSUE TITLE
Lesson 4 - loss blows up

### ISSUE THREAD

**Issue by BrianMiner** [roebius/deeplearning1_keras2#12]
Have you seen that sometimes the loss explodes when running the collaborative filtering?

**Comment by roebius** [comment]
No I haven't. If you are able to reproduce that behaviour it would be good to know the exact version of Cuda/CuDNN/Theano/Keras you are using.
Later this month I will release an update with some minor changes for using a more recent version of Keras.

**Comment by BrianMiner** [comment]
I am using tensorflow actually - theano is not installed.

tensorflow is 1.3.0
keras is 2.0.8

nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2016 NVIDIA Corporation
Built on Tue_Jan_10_13:22:03_CST_2017
Cuda compilation tools, release 8.0, V8.0.61

**Comment by BrianMiner** [comment]
It turns out to be a GPU issue. If I add

import os
os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"   # see issue #152
os.environ["CUDA_VISIBLE_DEVICES"] = ""

it works fine. It seems there are ongoing unresolved issues with GPU and keras?

https://github.com/keras-team/keras/issues/1244

**Comment by roebius** [comment]
So using the CPU your issue disappears. Not sure it is a Keras-only issue though.
I am using TF 1.3.0 with Cuda 8 and CuDNN 6. Are you using CuDNN 6 too, or could you give it a try?
(I am asking because sometimes I had weird behaviour on my workstation due to CuDNN version issues)

**Comment by BrianMiner** [comment]
I think this says I am using CUDNN 6

cat /usr/local/cuda/include/cudnn.h | grep CUDNN_MAJOR -A 2
#define CUDNN_MAJOR      6
#define CUDNN_MINOR      0
#define CUDNN_PATCHLEVEL 21
--
#define CUDNN_VERSION    (CUDNN_MAJOR * 1000 + CUDNN_MINOR * 100 + CUDNN_PATCHLEVEL)

#include "driver_types.h"

**Comment by roebius** [comment]
It is exactly my setup, so maybe it's Keras. Over the weekend I will update my repo with the changes for Keras 2.1.2. Maybe with Keras 2.1 the issue disappears.

**Comment by BrianMiner** [comment]
Just upgraded to 2.1.2 and no change.

**Comment by BrianMiner** [comment]
Upgraded Keras and all seems OK on my side now. Thanks for discussing!

**Comment by BrianMiner** [comment]
Oops what seemed to work was actually upgrading TF to 1.41. The upgrade of Keras did not seem to matter.

**Comment by roebius** [comment]
I am still on TF 1.3.0: if I discover anything that sheds some light on the issue I will let you know.
